hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ReplicationManager.java

```diff
@@ -448,6 +436,153 @@ protected ContainerHealthResult processContainer(ContainerInfo containerInfo,
+        .getContainerReplicas(containerID);
+    HddsProtos.LifeCycleState state = containerInfo.getState();
+    ContainerHealthResult health =
+        new ContainerHealthResult.HealthyResult(containerInfo);
+
+    if (state == HddsProtos.LifeCycleState.OPEN) {
+      if (!isOpenContainerHealthy(containerInfo, replicas)) {
+        report.incrementAndSample(
+            HealthState.OPEN_UNHEALTHY, containerID);
+        eventPublisher.fireEvent(SCMEvents.CLOSE_CONTAINER, containerID);
+        return new ContainerHealthResult.UnHealthyResult(containerInfo);
+      }
+      return health;
+    }
+
+    /*
+     * If the container is in CLOSING state, the replicas can either
+     * be in OPEN or in CLOSING state. In both of this cases
+     * we have to resend close container command to the datanodes.
+     */
+    if (state == HddsProtos.LifeCycleState.CLOSING) {
+      return handleClosingContainer(containerInfo, replicas);
+    }
+
+    //TODO: handle quosi_closed container for ratis after refactoring legacyRM
+
+    if (state == HddsProtos.LifeCycleState.CLOSED) {
+      List<ContainerReplica> unhealthyReplicas = replicas.stream()
+          .filter(r -> !compareState(containerInfo.getState(), r.getState()))
+          .collect(Collectors.toList());
+
+      if (unhealthyReplicas.size() > 0) {
+        return handleUnhealthyReplicas(containerInfo, unhealthyReplicas);
+      }
+    }
+
+    /*
+     * If container is under deleting and all it's replicas are deleted,
+     * then make the container as CLEANED,
+     * or resend the delete replica command if needed.
+     */
+    if (state == HddsProtos.LifeCycleState.DELETING) {
+      return handleDeletingContainer(containerInfo, replicas);
+    }
+
+    if (state == HddsProtos.LifeCycleState.DELETED) {
+      return health;
+    }
+
+    return null;
+  }
+
+  /**
+   * Handle the container which is in closing state.
+   *
+   * @param container ContainerInfo
+   * @param replicas Set of ContainerReplicas
+   */
+  private ContainerHealthResult handleClosingContainer(
+      final ContainerInfo container, final Set<ContainerReplica> replicas) {
+    ContainerHealthResult health =
+        new ContainerHealthResult.HealthyResult(container);
+    ContainerID containerID = container.containerID();
+    long term;
+    try {
+      term = scmContext.getTermOfLeader();
+    } catch (NotLeaderException nle) {
+      LOG.warn("Skip sending datanode command,"
+          + " since current SCM is not leader.", nle);
+      return health;
+    }
+
+    replicas.stream()
+        .filter(r -> r.getState() != ContainerReplicaProto.State.UNHEALTHY)
+        .forEach(r -> {
+          final CloseContainerCommand closeContainerCommand =
+              new CloseContainerCommand(
+                  containerID.getId(), container.getPipelineID(), false);
+          closeContainerCommand.setTerm(term);
+          closeContainerCommand.setEncodedToken(getContainerToken(containerID));
```

Comment:
Can you please extract a method for creating the close container command?  It's repeated in a few places.